### ISSUE TITLE
Crashes Comfy after Loaded state_dict from [....0F.ckpt]

### ISSUE THREAD

**Issue by gordingus** [kijai/ComfyUI-SUPIR#13]
Loading 1 new model
Diffusion using bf16
Encoder using bf16
Building a Downsample layer with 2 dims.
  --> settings are: 
 in-chn: 320, out-chn: 320, kernel-size: 3, stride: 2, padding: 1
constructing SpatialTransformer of depth 2 w/ 640 channels and 10 heads
constructing SpatialTransformer of depth 2 w/ 640 channels and 10 heads
Building a Downsample layer with 2 dims.
  --> settings are: 
 in-chn: 640, out-chn: 640, kernel-size: 3, stride: 2, padding: 1
constructing SpatialTransformer of depth 10 w/ 1280 channels and 20 heads
constructing SpatialTransformer of depth 10 w/ 1280 channels and 20 heads
constructing SpatialTransformer of depth 10 w/ 1280 channels and 20 heads
constructing SpatialTransformer of depth 10 w/ 1280 channels and 20 heads
constructing SpatialTransformer of depth 10 w/ 1280 channels and 20 heads
constructing SpatialTransformer of depth 10 w/ 1280 channels and 20 heads
constructing SpatialTransformer of depth 2 w/ 640 channels and 10 heads
constructing SpatialTransformer of depth 2 w/ 640 channels and 10 heads
constructing SpatialTransformer of depth 2 w/ 640 channels and 10 heads
Initialized embedder #0: FrozenCLIPEmbedder with 123060480 params. Trainable: False
Initialized embedder #1: FrozenOpenCLIPEmbedder2 with 694659841 params. Trainable: False
Initialized embedder #2: ConcatTimestepEmbedderND with 0 params. Trainable: False
Initialized embedder #3: ConcatTimestepEmbedderND with 0 params. Trainable: False
Initialized embedder #4: ConcatTimestepEmbedderND with 0 params. Trainable: False
making attention of type 'vanilla-xformers' with 512 in_channels
building MemoryEfficientAttnBlock with 512 in_channels...
Working with z of shape (1, 4, 32, 32) = 4096 dimensions.
making attention of type 'vanilla-xformers' with 512 in_channels
building MemoryEfficientAttnBlock with 512 in_channels...
Building a Downsample layer with 2 dims.
  --> settings are: 
 in-chn: 320, out-chn: 320, kernel-size: 3, stride: 2, padding: 1
constructing SpatialTransformer of depth 2 w/ 640 channels and 10 heads
constructing SpatialTransformer of depth 2 w/ 640 channels and 10 heads
Building a Downsample layer with 2 dims.
  --> settings are: 
 in-chn: 640, out-chn: 640, kernel-size: 3, stride: 2, padding: 1
constructing SpatialTransformer of depth 10 w/ 1280 channels and 20 heads
constructing SpatialTransformer of depth 10 w/ 1280 channels and 20 heads
constructing SpatialTransformer of depth 10 w/ 1280 channels and 20 heads
Loaded state_dict from [D:\Program Files\Visions of Chaos\Machine Learning Files\Text To Image\ComfyUI\ComfyUI\models\checkpoints\SUPIR\v0F.ckpt]

I was able to reproduce in a fresh install of portable as well as my current install through Visions of Chaos.
I've gotten it to work on smaller 512p images, but even 1.1x on a 1024 image didn't work.
0Q and 0F models both.
Freshly updated and reinstalled node from most recent commit.
3090, 24GB VRAM

**Comment by kijai** [comment]
How much system RAM do you have?

**Comment by gordingus** [comment]
64GB

**Comment by kijai** [comment]
Which model are you trying to load along with the SUPIR model? Something seems off with the models themselves, that part is very basic code wise, it's just loading the weights. I don't see it loading the SDXL model at all in that log.

**Comment by gordingus** [comment]
Full settings: 
![image](https://github.com/kijai/ComfyUI-SUPIR/assets/121201788/05511975-8362-4342-ab22-94d178d9aca6)

Would it load the SUPIR model before XL model? I'll have to try setting up a super basic workflow on a 512 and seeing what the log says

**Comment by kijai** [comment]
Yes SUPIR model is loaded first, should look like this:
![image](https://github.com/kijai/ComfyUI-SUPIR/assets/40791699/0fe1766b-960a-4735-9d55-5ec881040292)

**Comment by gordingus** [comment]
Yes, so it's dying just before loading state_dict of the XL model. Same thing happened on a 512p image with those same settings.

So if it prints the 0Q.ckpt, that means that step is successful and the XL model is causing the crash?

**Comment by kijai** [comment]
> Yes, so it's dying just before loading state_dict of the XL model. Same thing happened on a 512p image with those same settings.
> 
> So if it prints the 0Q.ckpt, that means that step is successful and the XL model is causing the crash?

Only settings of the node used at that point where it crashes for you is the model and dtype selections. You could try forcing the diffusion_type to fp16 and encoder to bf16, which is the default in the original, I've been using both bf16 as it seems to work well for me.

**Comment by gordingus** [comment]
Odd... When I switched to a different XL model, the 512 test image progressed to the end successfully and output the upscale.
Trying the same new model I did on the 1024 image crashed Comfy at the same place. Also tried bf16.

I'm stumped...

**Comment by gordingus** [comment]
Update: It appears to also happen when I switch models.
I ran the 512 once with Model A, outputs fine. Switch to Model B, crashes at the same point.
Restart Comfy, run Model B and it completes. Switch back to Model A, crashes at the same place.

**Comment by kijai** [comment]
Ah, so it's actually model switching that's possibly broken?

**Comment by gordingus** [comment]
Perhaps? That definitely did happen and then I got one XL model loaded and working through multiple images.

Though I think I may have stumbled on the real issue I was having: batch images

It definitely crashes when trying to get 1 image from Preview Chooser node > Batch 
![image](https://github.com/kijai/ComfyUI-SUPIR/assets/121201788/a59f2a5c-ecbf-4c9b-be8f-84d124c2b818)

Or this AnimateDiff workflow:
![image](https://github.com/kijai/ComfyUI-SUPIR/assets/121201788/e9157698-0ec3-45c6-83f8-1031f3a2deb4)

How did you get the sample video of the wolf batch processed?

**Comment by kijai** [comment]
Image list inputs are not supported, I simply made it handle batches one by one as default, as the assumption is that actual batch processing would be too heavy. So the wolf test was simply video loader straight to the node.

**Comment by gordingus** [comment]
Do you mean like this? I'm not getting multiple frames from the output, just one:
![image](https://github.com/kijai/ComfyUI-SUPIR/assets/121201788/227542e8-dc7a-4209-98b3-4ec853c798fb)

**Comment by kijai** [comment]
> Do you mean like this? I'm not getting multiple frames from the output, just one: ![image](https://private-user-images.githubusercontent.com/121201788/309207167-227542e8-dc7a-4209-98b3-4ec853c798fb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkyODQ3MzEsIm5iZiI6MTcwOTI4NDQzMSwicGF0aCI6Ii8xMjEyMDE3ODgvMzA5MjA3MTY3LTIyNzU0MmU4LWRjN2EtNDIwOS05OGIzLTRlYzg1M2M3OThmYi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMwMVQwOTEzNTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04NmQ5YWQ4MGQ3YmNjZmRiNDg3ODE3ZDFlNGNlNzYwNWNhMzU4MGYyZTJhZTNlZThkNzkwZjM1N2U3NmI1ZjQ0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.vZtD3-v_-gSIZt84FkBKiyOEgJ3Hf7oKeLDNMjHcQLo)

Ohhhh.. I'm sorry, at some point I have accidentally messed that up, should be fixed now.

**Comment by kijai** [comment]
Can you confirm if the model switching works now? Did small changes and at least I can't reproduce that issue anymore.

**Comment by gordingus** [comment]
After updating to c74b824 and some further testing:
1. Model Switching on the simplest workflow still crashes in the same way (Though it now prints "Attempting to load SDXL model" before crashing)
2. After fresh restart, without switching XL model, trying to use SUPIR in a wider workflow where an upscale would normally go. Crashes at same place despite no model switch. I'm even using the same model as the initial image generation.
Maybe it's because the initial generation model is already loaded and it can't handle a second instance? I tried converting XL model widget to input, but I can't connect the initial gen XL model to it because it's a COMBO type.
3. Had a problem with upscaled_image output going into Film VFI. I ended up getting a video to work by regular Save Image node, then importing the folder with those 24 frames and video combining. It looked quite good!
4. More of a question, how does the loading video frames as images to process one-by-one work? That's more what I meant by being able to batch process multiple images from AnimateDiff or Preview Chooser. 

So wondering if SUPIR node just needs to be used on its own in super simple workflow? I'd love it to be usable in my usual workflows because it's the best upscale I've seen. Thank you for your work on it!

**Comment by kijai** [comment]
> After updating to [c74b824](https://github.com/kijai/ComfyUI-SUPIR/commit/c74b8248a73352dc5bdc99496006e96321738f38) and some further testing:
> 
> 1. Model Switching on the simplest workflow still crashes in the same way (Though it now prints "Attempting to load SDXL model" before crashing)
> 2. After fresh restart, without switching XL model, trying to use SUPIR in a wider workflow where an upscale would normally go. Crashes at same place despite no model switch. I'm even using the same model as the initial image generation.
>    Maybe it's because the initial generation model is already loaded and it can't handle a second instance? I tried converting XL model widget to input, but I can't connect the initial gen XL model to it because it's a COMBO type.
> 3. Had a problem with upscaled_image output going into Film VFI. I ended up getting a video to work by regular Save Image node, then importing the folder with those 24 frames and video combining. It looked quite good!
> 4. More of a question, how does the loading video frames as images to process one-by-one work? That's more what I meant by being able to batch process multiple images from AnimateDiff or Preview Chooser.
> 
> So wondering if SUPIR node just needs to be used on its own in super simple workflow? I'd love it to be usable in my usual workflows because it's the best upscale I've seen. Thank you for your work on it!

Odd that I can't reproduce that at all, but I'm also only using it standalone as it's just so memory heavy. Could also be RAM running out, the clip model seems to take huge amount when loading. 

Currently it's not using the comfy model management properly, so incorporating to other workflows is rough. I can look into that at some point and at minimum incorporate other model unloading to free memory better. In fact I'll add the bare minimum of unload_all_models to the beginning, no harm in it in this case, may help or may not.

Saving the frames one by one is probably the best approach overall, as it's so slow and only benefit of batching is convenience as the frames are processed one by one too.

One way to do that is a setup like this, where every queue increments the primitive node and thus you can just queue x amount of times and it runs through the frames:
![image](https://github.com/kijai/ComfyUI-SUPIR/assets/40791699/ef86944c-e159-4f60-b148-c0fa16f88a9e)

**Comment by gordingus** [comment]
Unload models doesn't seem to have made a difference. Also I tried disabling all custom nodes in case it was rgthree or something interfering with a model switch. No dice, model switch in this simplest workflow crashes at the same point:

![image](https://github.com/kijai/ComfyUI-SUPIR/assets/121201788/bd3c1bc8-4be1-4d32-ab53-5701f47c29c7)

**Comment by kijai** [comment]
> Unload models doesn't seem to have made a difference. Also I tried disabling all custom nodes in case it was rgthree or something interfering with a model switch. No dice, model switch in this simplest workflow crashes at the same point:
> 
> ![image](https://private-user-images.githubusercontent.com/121201788/309384773-bd3c1bc8-4be1-4d32-ab53-5701f47c29c7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkzMjg4NTQsIm5iZiI6MTcwOTMyODU1NCwicGF0aCI6Ii8xMjEyMDE3ODgvMzA5Mzg0NzczLWJkM2MxYmM4LTRiZTEtNGQzMi1hYjUzLTU3MDFmNDdjMjljNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMwMVQyMTI5MTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wZWE0MTcwMWViMDhlMGZjNTgwNzZmNmUyNjk0MWY3ZDU2OGVlZjM1ZjcxYzhiMmZiNzllYTEyMTZlYmUyNGI5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.xeLIxgabn6U6S8WFKHmzSq7v2Ml6qzUFo2TC1C5f-GU)

Added cache cleaning additionally when switching models, for me it does free all memory just fine... 
In better news, thanks to a tip from Fannovel, the model loading is thrice as fast now!

**Comment by gordingus** [comment]
Hey, that fixed model switching! I switched to multiple models using the above simple workflow with no crashes. Thanks to Fannovel and you 🎉

Still crashes in a wider workflow at the same point, but it's definitely progress. Might even close this issue? Though I'd want to add some of the features I've talked about in this thread to a wishlist if you're interested?

**Comment by kijai** [comment]
Yeah I believe this particular issue is resolved, feel free to make a new one for that stuff though!
